compiler/rustc_lint/src/builtin.rs

```diff
@@ -3064,3 +3065,95 @@ impl EarlyLintPass for SpecialModuleName {
         }
     }
 }
+
+declare_lint! {
+    /// The `self_type_conversion` lint detects when a call to `.into()` does not have any effect.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// fn main() {
+    ///     let () = ().into();
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    pub SELF_TYPE_CONVERSION,
+    Deny,
+    "",
+}
+
+pub struct SelfTypeConversion<'tcx> {
+    ignored_types: Vec<Ty<'tcx>>,
+}
+
+impl_lint_pass!(SelfTypeConversion<'_> => [SELF_TYPE_CONVERSION]);
+
+impl SelfTypeConversion<'_> {
+    pub fn new() -> Self {
+        Self { ignored_types: vec![] }
+    }
+}
+
+impl<'tcx> LateLintPass<'tcx> for SelfTypeConversion<'tcx> {
+    fn check_item_post(&mut self, cx: &LateContext<'tcx>, item: &hir::Item<'_>) {
+        let hir::ItemKind::Use(path, kind) = item.kind else { return };
+        tracing::info!("{:#?}", item);
+        tracing::info!(?path, ?kind);
+        for res in &path.res {
+            let Res::Def(DefKind::TyAlias, def_id) = res else { continue };
+            let ty = cx.tcx.type_of(def_id).instantiate_identity();
+            let name = cx.tcx.item_name(*def_id);
+            // println!("{ty:?} {name:?}");
+            self.ignored_types.push(ty);
+            for stripped in cx.tcx.stripped_cfg_items(def_id.krate) {
+                if stripped.name.name == name {
+                    tracing::info!("{name:#?}");
+                }
+            }
+        }
+    }
```

Comment:
This is not the right way to do this check, but it was the fastest way of getting a semi-accurate discard of types that have `type` aliases in scope to somewhat handle platform `cfg` cases.